src/frontend/src/optimizer/backfill_order_strategy.rs

```diff
@@ -0,0 +1,388 @@
+// Copyright 2025 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_pb::stream_plan::BackfillOrderStrategy as PbBackfillOrderStrategy;
+use risingwave_sqlparser::ast::BackfillOrderStrategy;
+
+use crate::PlanRef;
+use crate::error::Result;
+use crate::optimizer::backfill_order_strategy::auto::plan_auto_strategy;
+use crate::optimizer::backfill_order_strategy::fixed::plan_fixed_strategy;
+use crate::session::SessionImpl;
+
+mod auto {
+    use std::collections::{HashMap, HashSet};
+
+    use risingwave_common::catalog::ObjectId;
+    use risingwave_pb::common::Uint32Vector;
+    use risingwave_pb::stream_plan::BackfillOrderFixed;
+    use risingwave_pb::stream_plan::backfill_order_strategy::Strategy as PbStrategy;
+
+    use crate::PlanRef;
+    use crate::optimizer::PlanNodeType;
+    use crate::optimizer::backfill_order_strategy::common::has_cycle;
+    use crate::session::SessionImpl;
+
+    #[derive(Debug)]
+    enum BackfillTreeNode {
+        Join {
+            lhs: Box<BackfillTreeNode>,
+            rhs: Box<BackfillTreeNode>,
+        },
```

Comment:
I think the join should have a direction information like `Left` and `Right`, because for left outer join and right outer join, the backfill order should be different.

Reply:
Will support it in a separate PR. https://github.com/risingwavelabs/risingwave/issues/20861.